book/src/pil/types.md

```diff
@@ -9,13 +9,75 @@ The powdr-pil language has the following types:
 - tuple
 - array
 - function type
-- expression
-- constraint
+- `expr` (expression)
+- `constr` (constraint)
+- `!` ("bottom" or "unreachable" type)
+
+> The `col` type is special in that it is not a proper type, but can be used for declaring symbols.
+> See [Types of Symbols](#types-of-symbols) for details.
+
+Powdr-pil does perform a Hindley-Milner-kind type checking. After type checking, for all symbols
+without an explicit type, their type has to be uniquely determined. If you want a symbol to have
+a generic type, that generic type has to be explicitly specified.
+
+Declarations of generic symbols can constraint type variables to satisfy certain bounds.
+Currently, only built-in bounds or traits are supported (see the next section).
+
+Literal numbers do not have a specific type, they can be either `int`, `fe` or `expr` (the types that
+implement the `FromLiteral` trait), and their type can also stay generic until evaluation.
+
+### Example
+
+The following snippet defines a function that takes a value of a generic type and returns the value incremented by one.
+The type bounds on the generic type are `FromLiteral` and `Add`. The type checker will complain if we do not specify
+the type bounds. The bound `Add` is required because we use the `+` operator in the function and `FromLiteral` is needed
+because we use the literal `1` as a value of that type.
+
+```rust
+let<T: FromLiteral + Add> add_one: T -> T = |i| i + 1;
+```
+
+## Types of Symbols
```

Comment:
Maybe call this section `Declaring and referencing columns` or sth in those lines?